File: src/lowlevel/server/icefsoperators/icefsSetXattr.go
/*
 * @Author: Tan90degrees tangentninetydegrees@gmail.com
 * @Date: 2023-03-11 07:18:32
 * @LastEditors: Tan90degrees tangentninetydegrees@gmail.com
 * @LastEditTime: 2023-04-18 14:42:37
 * @FilePath: /icefs/src/lowlevel/server/icefsoperators/icefsSetXattr.go
 * @Description:
 *
 * Copyright (C) 2023 Tan90degrees <tangentninetydegrees@gmail.com>.
 */
package icefsoperators

import (
	"context"
	"fmt"
	"icefs-server/icefserror"
	pb "icefs-server/icefsgrpc"
	"icefs-server/icefsthrift"
	"syscall"
)

func (s *IcefsServer) doIcefsSetXattr(fakeInode uint64, name string, value string, flags int32) (status int32) {
	var procName string
	s.inodeCacheLock.RLock()
	inode := s.getIcefsInode(fakeInode)
	if inode == nil {
		s.inodeCacheLock.RUnlock()
		status = icefserror.ICEFS_BUG_ERR
		return
	}
	inode.inodeLock.RLock()
	s.inodeCacheLock.RUnlock()
	procName = fmt.Sprintf("/proc/self/fd/%v", inode.fd)
	inode.inodeLock.RUnlock()
	status = icefserror.IcefsStdErrno(syscall.Setxattr(procName, name, []byte(value), int(flags)))
	return
}

func (s *IcefsGRpcServer) DoIcefsSetXattr(ctx context.Context, req *pb.IcefsSetXattrReq) (*pb.IcefsSetXattrRes, error) {
	var res pb.IcefsSetXattrRes

	res.Status = s.server.doIcefsSetXattr(req.Inode, req.Name, req.Value, req.Flags)

	return &res, nil
}

func (s *IcefsThriftServer) DoIcefsSetXattr(ctx context.Context, req *icefsthrift.IcefsSetXattrReq) (*icefsthrift.IcefsSetXattrRes, error) {
	var res icefsthrift.IcefsSetXattrRes

	res.Status = s.server.doIcefsSetXattr(uint64(req.Inode), req.Name, req.Value, req.Flags)

	return &res, nil
}

File: src/lowlevel/client/icefsoperators/icefsSetXattr.cpp
/*
 * @Author: Tan90degrees tangentninetydegrees@gmail.com
 * @Date: 2023-03-30 04:19:29
 * @LastEditors: Tan90degrees tangentninetydegrees@gmail.com
 * @LastEditTime: 2023-04-21 02:41:04
 * @FilePath: /icefs/src/lowlevel/client/icefsoperators/icefsSetXattr.cpp
 * @Description:
 *
 * Copyright (C) 2023 Tan90degrees <tangentninetydegrees@gmail.com>.
 */
#include <stdio.h>

#include <string>

#include "icefsClient.hpp"
#include "icefsClientUtils.hpp"

void IcefsClient::DoIcefsSetXattr(fuse_req_t fuseReq, fuse_ino_t inode,
                                  const char *name, const char *value,
                                  size_t size, int flags) {
  ICEFS_PR_FUNCTION;

  switch (this->clientConfig.linkType) {
    case ICEFS_LINK_USE_GRPC: {
      icefsgrpc::IcefsSetXattrReq req;
      icefsgrpc::IcefsSetXattrRes res;
      grpc::ClientContext ctx;

      req.set_inode(inode);
      req.set_name(name);
      req.set_value(value);
      req.set_flags(flags);

      grpc::Status status = gRpcClient->DoIcefsSetXattr(&ctx, req, &res);
      if (status.ok() && !res.status()) {
        fuse_reply_err(fuseReq, ICEFS_EOK);
      } else {
        fuse_reply_err(fuseReq, res.status() ? res.status() : EIO);
        ICEFS_PR_ERR_STATUS;
      }
      break;
    }

    case ICEFS_LINK_USE_THRIFT: {
      icefsthrift::IcefsSetXattrReq req;
      icefsthrift::IcefsSetXattrRes res;

      req.__set_inode(inode);
      req.__set_name(name);
      req.__set_value(value);
      req.__set_flags(flags);

      icefsThriftConn *thriftConn = thriftClientPool->GetIcefsThriftConn();
      thriftConn->thriftClient->DoIcefsSetXattr(res, req);
      thriftClientPool->PutIcefsThriftConn(thriftConn);
      fuse_reply_err(fuseReq, res.status);
      break;
    }

    default:
      fuse_reply_err(fuseReq, EIO);
      break;
  }
}
